lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,462 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This file includes methods to manage TLS certificates within the Vault charms."""
+
+import logging
+import socket
+from enum import Enum, auto
+from signal import SIGHUP
+from typing import List, Optional, TextIO, Tuple
+
+from charms.certificate_transfer_interface.v0.certificate_transfer import (
+    CertificateTransferProvides,
+)
+from charms.tls_certificates_interface.v2.tls_certificates import (
+    TLSCertificatesRequiresV2,
+    generate_ca,
+    generate_certificate,
+    generate_csr,
+    generate_private_key,
+)
+from ops import EventBase, Object, RelationBrokenEvent, SecretNotFoundError
+from ops.pebble import APIError, PathError
+
+from exceptions import PeerSecretError, VaultCertsError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "61b41a053d9847ce8a14eb02197d12cb"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class TLSAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend TLS to all log lines."""
+
+    def process(self, msg, kwargs):
+        """Decides the format for the prepended text."""
+        return "[%s]%s" % (self.extra, msg), kwargs
+
+
+tlslogger = TLSAdapter(logger, "TLS")  # type: ignore[arg-type]
```

Comment:
This was how the example in the [python docs](https://docs.python.org/3/howto/logging-cookbook.html#adding-contextual-information-to-your-logging-output) were given, but I can hardcore the string.

However, mypy will still complain about the second type:
```
error: Missing positional argument "extra" in call to "TLSAdapter"  [call-arg]
    tlslogger = TLSAdapter(logger)
```
even though a default value has been provided.

Reply:
Not sure what you mean in reference to the docs. The docs show it used in two ways

Without the extras:

```
class StyleAdapter(logging.LoggerAdapter):
    def __init__(self, logger, extra=None):
        super().__init__(logger, extra or {})

    def log(self, level, msg, /, *args, **kwargs):
        if self.isEnabledFor(level):
            msg, kwargs = self.process(msg, kwargs)
            self.logger._log(level, Message(msg, args), (), **kwargs)

logger = StyleAdapter(logging.getLogger(__name__))
```

Or with the extras:

```
class CustomAdapter(logging.LoggerAdapter):
    """
    This example adapter expects the passed in dict-like object to have a
    'connid' key, whose value in brackets is prepended to the log message.
    """
    def process(self, msg, kwargs):
        return '[%s] %s' % (self.extra['connid'], msg), kwargs

logger = logging.getLogger(__name__)
adapter = CustomAdapter(logger, {'connid': some_conn_id})
```

But, it looks like extras wasn't made optional until 3.10, so if we want to support lower than 3.10, we need to pass an empty dict `{}`, or make it optional in the `TLSAdapter` ourselves. Regardless, the type string (`"TLS"`) is wrong.

```
tlslogger = TLSAdapter(logger, {})
```

Not sure where you're getting the idea there needs to be a `subject`?

Reply:
Ahh, I see. You've used the extra in the `process` call. I missed that at first. It shouldn't be a string, so at least that needs to be updated. But I would just make process add TLS directly and not use the convoluted extras because we don't need it to be configurable.

```
class TLSAdapter(logging.LoggerAdapter):
    """Adapter for the logger to prepend TLS to all log lines."""

    def process(self, msg, kwargs):
        """Decides the format for the prepended text."""
        return "[TLS] " + msg, kwargs
```

Either is fine, just not using extras as a string :sweat_smile:

Reply:
Done